### PR TITLE
MM-52154 pagination for boards APIs

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
This PR adds pagination to the following REST API's:
- GetBlocks
- GetBoards
- GetCards
- GetUserCategoryBoards
- TeamBoardsInsights
- UserBoardsInsights
- GetMembersForBoard
- GetSubscriptions
- GetMyMemberships

App layer and store layer APIs were also modified to add pagination, and any dependent code updated to use pagination.  Some TODOs were added to paginate the few remaining APIs that could not be paginated without significant refactoring.

**NOTE:**  the mattermostauthlayer store APIs have no unit tests and adding them are out of scope for this PR as that is a sizable project.  Unfortunately that means the most complex parts of this PR have no unit tests and will require comprehensive testing.

**NOTE 2:** currently the pagination parameters are optional in the affected REST APIs pending changes to webapp to supply the params and paginate the results.  Once the webapp is updated, one line of code can be changed to make the params mandatory:  `server/boards/api/api/go:162`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-52154

fixes https://github.com/mattermost/focalboard/issues/4406

#### Release Note
```release-note
Pagination added to all Boards REST APIs that return unbounded result sets.
```

### CODE REVIEW

File: server/boards/services/store/mattermostauthlayer/mattermostauthlayer.go [hunk 1]
```diff
@@ -344,14 +344,23 @@ func (s *MattermostAuthLayer) GetUsersByTeam(teamID string, asGuestID string, sh
...
-
 		boardsIDs := []string{}
-		for _, board := range boards {
-			boardsIDs = append(boardsIDs, board.ID)
+		page := 0
+		for ; true; page++ {
+			boards, err := s.GetBoardsForUserAndTeam(asGuestID, teamID, model.QueryBoardOptions{
+				IncludePublicBoards: false,
+				Page:                page,
+				PerPage:             100,
```
Comment (mgdelacroix) [inline server/boards/services/store/mattermostauthlayer/mattermostauthlayer.go:353]: Same constant suggestion here and a few lines below

**Review by mgdelacroix — COMMENTED** [review]
Amazing work! :sparkles:

**Review by sbishel — COMMENTED** [review]
Wow. Nice, work. I'll wait for final changes before approving. But  LGTM.

There are some todos, we should create tickets for, so we can prioritize. Let me know if you want to discuss the union queries refactor.

File: server/boards/app/boards.go [hunk 2]
```diff
@@ -419,30 +445,39 @@ func (a *App) PatchBoard(patch *model.BoardPatch, boardID, userID string) (*mode
...
+					a.wsAdapter.BroadcastMemberDelete(updatedBoard.TeamID, boardID, oldMemberID)
 				}
 			}
 		}
 
 		if patch.Type != nil && isTemplate {
-			members, err := a.GetMembersForBoard(updatedBoard.ID)
-			if err != nil {
-				a.logger.Error("Unable to get the board members", mlog.Err(err))
+			page := 0
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/boards/app/boards.go:432]: The `broadcastTeamUsers` API calls into `GetTeamMembers` and it is better to call that once per 50 members than for each member individually.  Frankly, `GetTeamUsers` should be paginated as well, but channel APIs are out of scope.

**Comment by wiggin77** [comment]
> I was thinking about this PR and realized that the two mattermostauthlayer APIs that have a `UNION DISTINCT` won't remove duplicates (explicit and implicit users can overlap) because each query has a different value hard-coded for `synthetic` and `roles`. Either the duplicates will have to be resolved at the client (ugh) or I'll have to refactor these API's. 

I found a relatively simple solution to the above.  Fixed.

**Review by sbishel — APPROVED** [review]
Awesome!

File: server/boards/app/boards.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply server/boards/app/boards.go:433]: Makes total sense

**Review by mgdelacroix — APPROVED** [review]

**Comment by wiggin77** [comment]
@sbishel I am assigning this PR to you to keep it on your radar.  I understand it may get merged in the future or closed.

**Comment by wiggin77** [comment]
Closing as this was done against mono-repo.

**Pull request closed without merging**